<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Комментарий-ревьюера." data-toc-modified-id="Комментарий-ревьюера.-0.1"><span class="toc-item-num">0.1&nbsp;&nbsp;</span>Комментарий ревьюера.</a></span></li></ul></li><li><span><a href="#Знакомство-с-данными" data-toc-modified-id="Знакомство-с-данными-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Знакомство с данными</a></span><ul class="toc-item"><li><span><a href="#Исследование-таблицы-books" data-toc-modified-id="Исследование-таблицы-books-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Исследование таблицы <code>books</code></a></span></li><li><span><a href="#Исследование-таблицы-authors" data-toc-modified-id="Исследование-таблицы-authors-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Исследование таблицы <code>authors</code></a></span></li><li><span><a href="#Исследование-таблицы-publishers" data-toc-modified-id="Исследование-таблицы-publishers-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Исследование таблицы <code>publishers</code></a></span></li><li><span><a href="#Исследование-таблицы-ratings" data-toc-modified-id="Исследование-таблицы-ratings-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Исследование таблицы <code>ratings</code></a></span></li><li><span><a href="#Исследование-таблицы-reviews" data-toc-modified-id="Исследование-таблицы-reviews-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Исследование таблицы <code>reviews</code></a></span></li><li><span><a href="#Вывод-по-исследованию-данных" data-toc-modified-id="Вывод-по-исследованию-данных-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>Вывод по исследованию данных</a></span></li></ul></li><li><span><a href="#Анализ-данных" data-toc-modified-id="Анализ-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Анализ данных</a></span><ul class="toc-item"><li><span><a href="#Количество-книг,-которые-вышли-после-1-января-2000-года" data-toc-modified-id="Количество-книг,-которые-вышли-после-1-января-2000-года-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Количество книг, которые вышли после 1 января 2000 года</a></span></li><li><span><a href="#Количество-обзоров-и-средняя-оценка-для-каждой-книги" data-toc-modified-id="Количество-обзоров-и-средняя-оценка-для-каждой-книги-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Количество обзоров и средняя оценка для каждой книги</a></span></li><li><span><a href="#Издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц" data-toc-modified-id="Издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Издательство, которое выпустило наибольшее число книг толще 50 страниц</a></span></li><li><span><a href="#Автор-с-самой-высокой-средней-оценкой-книг" data-toc-modified-id="Автор-с-самой-высокой-средней-оценкой-книг-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Автор с самой высокой средней оценкой книг</a></span></li><li><span><a href="#Среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок" data-toc-modified-id="Среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Среднее количество обзоров от пользователей, которые поставили больше 50 оценок</a></span></li><li><span><a href="#Вывод-по-анализу-данных" data-toc-modified-id="Вывод-по-анализу-данных-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Вывод по анализу данных</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Общий вывод</a></span><ul class="toc-item"><li><span><a href="#Проделанная-работа" data-toc-modified-id="Проделанная-работа-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Проделанная работа</a></span></li><li><span><a href="#Общие-выводы" data-toc-modified-id="Общие-выводы-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Общие выводы</a></span></li><li><span><a href="#Рекомендации" data-toc-modified-id="Рекомендации-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Рекомендации</a></span></li></ul></li></ul></div>

# Анализ базы данных с помощью SQL-запросов

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Наша компания решила быть на волне и купила крупный сервис для чтения книг по подписке.В базе данных содержится информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Цель исследования**: провести анализ базы данных, используюя язык запросов SQL.

**Описание данных**

1. Таблица `books` содержит данные о книгах:
- _book_id_ — идентификатор книги;
- _author_id_ — идентификатор автора;
- _title_ — название книги;
- _num_pages_ — количество страниц;
- _publication_date_ — дата публикации книги;
- _publisher_id_ — идентификатор издателя.
2. Таблица `authors` содержит данные об авторах:
- _author_id_ — идентификатор автора;
- _author_ — имя автора.
3. Таблица `publishers` содержит данные об издательствах:
- _publisher_id_ — идентификатор издательства;
- _publisher_ — название издательства;
4. Таблица `ratings` содержит данные о пользовательских оценках книг:
- _rating_id_ — идентификатор оценки;
- _book_id_ — идентификатор книги;
- _username_ — имя пользователя, оставившего оценку;
- _rating_ — оценка книги.
5. Таблица `reviews` содержит данные о пользовательских обзорах:
- _review_id_ — идентификатор обзора;
- _book_id_ — идентификатор книги;
- _username_ — имя автора обзора;
- _text_ — текст обзора.

## Знакомство с данными

Перед началом работы с данными импортируем все необходимые библиотеки:

In [1]:
import pandas as pd
from sqlalchemy import create_engine

Зададим настройки отображения:

In [2]:
pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', '{:.2f}'.format)
pd.set_option("max_colwidth", 500)

Установим параметры подключения к базе данных:

In [3]:
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])

Сохраним коннектор в переменной _engine_:

In [4]:
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

Напишем функцию для вывода на экран первых 10 строк таблицы:

In [5]:
def table_info(table):
    """Display information about the table"""
    
    print('Первые 10 строк таблицы:')
    display(pd.io.sql.read_sql('''SELECT *
                                  FROM {} 
                                  LIMIT 10'''.format(table), con = engine))

### Исследование таблицы `books`

С помощью функции _table_info_ выведем на экран первые 10 строк таблицы `books`:

In [6]:
table_info('books')

Первые 10 строк таблицы:


book_id  author_id                                                  title  \
0        1        546                                           'Salem's Lot   
1        2        465                     1 000 Places to See Before You Die   
2        3        407    13 Little Blue Envelopes (Little Blue Envelope  #1)   
3        4         82  1491: New Revelations of the Americas Before Columbus   
4        5        125                                                   1776   
5        6        257                   1st to Die (Women's Murder Club  #1)   
6        7        258                   2nd Chance (Women's Murder Club  #2)   
7        8        260                  4th of July (Women's Murder Club  #4)   
8        9        563                                       A Beautiful Mind   
9       10        445                                     A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

### Исследование таблицы `authors`

С помощью функции _table_info_ выведем на экран первые 10 строк таблицы `authors`:

In [7]:
table_info('authors')

Первые 10 строк таблицы:


author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

### Исследование таблицы `publishers`

С помощью функции _table_info_ выведем на экран первые 10 строк таблицы `publishers`:

In [8]:
table_info('publishers')

Первые 10 строк таблицы:


publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

### Исследование таблицы `ratings`

С помощью функции _table_info_ выведем на экран первые 10 строк таблицы `ratings`:

In [9]:
table_info('ratings')

Первые 10 строк таблицы:


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

### Исследование таблицы `reviews`

С помощью функции _table_info_ выведем на экран первые 10 строк таблицы `reviews`:

In [10]:
table_info('reviews')

Первые 10 строк таблицы:


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                                                                                               text  
0                                        Mention society tell send professor analysis. Over provide race technology continue these.  
1                                                 Foot glass pretty audience hit themselves. Among admit investment argue security.  
2  Listen treat keep worry. Miss husband tax but person sport treatment industry. Kitchen decision deep the. Social party body the.  
3          Finally month interesting blue could nature cultural bit. Prepare beat finish grow that smile teach. Dream me play near.  
4                  Nation purpose heavy give wait song will. List dinner another whole positive radio fast. Music staff many green.  
5                                                                                          Analysis no several cause international.  
6                              One there cost another. Say type save. With personal where occur direction foot. Half sit role want.  
7                                                                 Within enough mother. There at system full recent play recognize.  
8                                  Thank now focus realize economy focus fly. Item step fine war. Western service history anything.  
9            Game push lot reduce where remember. Including song travel hair foot. Director always myself past story chance reason.

### Вывод по исследованию данных

В ходе исследования предоставленных данных, была проделана следующая работа:
- Было установлено подключение к СУБД PostgreSQL;
- Проведено знакомство с данными, содержащимися в таблицах: `books`, `authors`, `publishers`, `ratings` и `reviews`.

## Анализ данных

### Количество книг, которые вышли после 1 января 2000 года

С помощью языка запросов SQL определим количество книг, которые вышли после 1 января 2000 года:

In [11]:
query = '''SELECT COUNT(book_id) as count_books
           FROM books
           WHERE publication_date >= '2000-01-01';
        '''
pd.io.sql.read_sql(query, con = engine)

count_books
0          821

Вывод:
- Количество книг, которые вышли после 1 января 2000 года - **821** штука.

### Количество обзоров и средняя оценка для каждой книги

С помощью языка запросов SQL определим количество обзоров и среднюю оценку для каждой книги:

In [12]:
query = '''SELECT b.title AS title,
                  COUNT(DISTINCT rev.review_id) as count_reviews,
                  AVG(rat.rating) as avg_rating
           FROM books AS b
           LEFT JOIN reviews AS rev ON b.book_id = rev.book_id
           LEFT JOIN ratings AS rat ON b.book_id = rat.book_id
           GROUP BY title
           ORDER BY count_reviews DESC;
        '''
pd.io.sql.read_sql(query, con = engine)

title  \
0                                           Memoirs of a Geisha   
1                                       Twilight (Twilight  #1)   
2    Harry Potter and the Chamber of Secrets (Harry Potter  #2)   
3                                              The Glass Castle   
4                                               Eat  Pray  Love   
..                                                          ...   
994             Anne Rice's The Vampire Lestat: A Graphic Novel   
995                                     The Natural Way to Draw   
996            The Cat in the Hat and Other Dr. Seuss Favorites   
997                                   Essential Tales and Poems   
998        Disney's Beauty and the Beast (A Little Golden Book)   

     count_reviews  avg_rating  
0                8        4.14  
1                7        3.66  
2                6        4.29  
3                6        4.21  
4                6        3.40  
..             ...         ...  
994              0        3.67  
995              0        3.00  
996              0        5.00  
997              0        4.00  
998              0        4.00  

[999 rows x 3 columns]

Выводы:
- Несколько книг имеют максимальную среднюю оценку **5 баллов**;
- Средняя оценка не зависит от количества обзоров.

### Издательство, которое выпустило наибольшее число книг толще 50 страниц

С помощью языка запросов SQL определим издательство, которое выпустило наибольшее число книг толще 50 страниц, таким образом исключив из анализа брошюры:

In [13]:
query = '''SELECT p.publisher AS publisher,
                  COUNT(b.book_id) as count_books
           FROM publishers AS p
           LEFT JOIN books AS b ON p.publisher_id = b.publisher_id
           WHERE num_pages > 50
           GROUP BY publisher
           ORDER BY count_books DESC
           LIMIT 1;
        '''
pd.io.sql.read_sql(query, con = engine)

publisher  count_books
0  Penguin Books           42

Вывод:
- Издательство **Penguin Books** выпустило наибольшее число книг толще 50 страниц - **42 книги**, не включая брошюры.

### Автор с самой высокой средней оценкой книг

С помощью языка запросов SQL определим автора с самой высокой средней оценкой книг, при этом, учитывая только книги с 50 и более оценками:

In [14]:
query = '''SELECT author, 
                  AVG(avg_rating) AS avg_rating_author
           FROM (SELECT author_id, 
                        title, 
                        AVG(rating) AS avg_rating
                 FROM books AS b
                 LEFT JOIN ratings AS rat ON b.book_id = rat.book_id 
                 GROUP BY title, author_id
                 HAVING COUNT(rating_id) >= 50) AS avg
           LEFT JOIN authors AS a ON avg.author_id = a.author_id
           GROUP BY author
           ORDER BY avg_rating_author DESC
           LIMIT 1;
        '''
pd.io.sql.read_sql(query, con = engine)

author  avg_rating_author
0  J.K. Rowling/Mary GrandPré               4.28

Вывод:
- Автор с самой высокой средней оценкой книг - **J.K. Rowling/Mary GrandPré** со средней оценкой **4.28**.

### Среднее количество обзоров от пользователей, которые поставили больше 50 оценок

С помощью языка запросов SQL определим среднее количество обзоров от пользователей, которые поставили больше 50 оценок:

In [15]:
query = '''SELECT ROUND(AVG(count_reviews)) AS avg_reviews
           FROM (SELECT COUNT(review_id) AS count_reviews
                 FROM (SELECT username
                       FROM ratings AS rat
                       GROUP BY username
                       HAVING COUNT(rating) > 50) AS rat
                 LEFT JOIN reviews AS rev ON rat.username = rev.username
                 GROUP BY rat.username) AS rev;
        '''
pd.io.sql.read_sql(query, con = engine)

avg_reviews
0        24.00

Вывод:
- Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - **24 штуки**.

### Вывод по анализу данных

По проведенному анализу можно сделать следующие выводы:
- Количество книг, которые вышли после 1 января 2000 года - **821 штука**;
- Несколько книг имеют максимальную среднюю оценку **5 баллов**;
- Средняя оценка **не зависит** от количества обзоров;
- Издательство **Penguin Books** выпустило наибольшее число книг толще 50 страниц - **42 книги**, не включая брошюры;
- Автор с самой высокой средней оценкой книг - **J.K. Rowling/Mary GrandPré** со средней оценкой **4.28**;
- Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - **24 штуки**.

## Общий вывод

### Проделанная работа

В ходе проведения анализа была проделана следующая работа:

**1. Знакомство с данными, в ходе которого:**
- Было установлено подключение к СУБД PostgreSQL;
- Исследованы данные, содержащиеся в таблицах: `books`, `authors`, `publishers`, `ratings` и `reviews`.

**2. Анализ данных с помощью языка запросов SQL, в ходе которого:**
- Определено количество книг, которые вышли после 1 января 2000 года;
- Определены количество обзоров и средняя оценка для каждой книги;
- Определено издательство, которое выпустило наибольшее число книг толще 50 страниц;
- Определен автор с самой высокой средней оценкой книг;
- Определено среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

### Общие выводы

В ходе проведения анализа были получены следующие выводы:
- Количество книг, которые вышли после 1 января 2000 года - **821 штука**;
- Несколько книг имеют максимальную среднюю оценку **5 баллов**;
- Средняя оценка **не зависит** от количества обзоров;
- Издательство **Penguin Books** выпустило наибольшее число книг толще 50 страниц - **42 книги**, не включая брошюры;
- Автор с самой высокой средней оценкой книг - **J.K. Rowling/Mary GrandPré** со средней оценкой **4.28**;
- Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - **24 штуки**.

### Рекомендации

По результатам проведенного анализа можно дать следующие рекоммендации:
- Продолжать тесное сотрудничество с издательством **Penguin Books**;
- Обратить особое внимание на книги **J.K. Rowling/Mary GrandPré**, которые интересны читателям.